pkg/volume/aws_ebs/aws_util.go

```diff
@@ -112,62 +148,76 @@ func (util *AWSDiskUtil) CreateVolume(c *awsElasticBlockStoreProvisioner, node *
-			return "", 0, nil, "", fmt.Errorf("invalid option %q for volume plugin %s", k, c.plugin.GetPluginName())
+			return nil, fmt.Errorf("invalid option %q for volume plugin %s", k, pluginName)
 		}
 	}
 
 	if volumeOptions.ZonePresent && volumeOptions.ZonesPresent {
-		return "", 0, nil, "", fmt.Errorf("both zone and zones StorageClass parameters must not be used at the same time")
+		return nil, fmt.Errorf("both zone and zones StorageClass parameters must not be used at the same time")
 	}
 
-	if utilfeature.DefaultFeatureGate.Enabled(features.DynamicProvisioningScheduling) && node != nil {
-		if node.ObjectMeta.Labels == nil {
-			return "", 0, nil, "", fmt.Errorf("labels for node missing")
-		}
-		z, ok := node.ObjectMeta.Labels[kubeletapis.LabelZoneFailureDomain]
-		if !ok {
-			return "", 0, nil, "", fmt.Errorf("%s Label for node missing", kubeletapis.LabelZoneFailureDomain)
-		}
-
-		if volumeOptions.ZonePresent || volumeOptions.ZonesPresent {
-			return "", 0, nil, "", fmt.Errorf("Zone/Zones cannot be specified in StorageClass if VolumeBindingMode set to WaitForFirstConsumer.")
-		}
-
-		volumeOptions.ZonePresent = true
-		volumeOptions.AvailabilityZone = z
-	}
-
-	// TODO: implement PVC.Selector parsing
-	if c.options.PVC.Spec.Selector != nil {
-		return "", 0, nil, "", fmt.Errorf("claim.Spec.Selector is not supported for dynamic provisioning on AWS")
-	}
+	if utilfeature.DefaultFeatureGate.Enabled(features.DynamicProvisioningScheduling) {
+		if node != nil {
+			// if node set, zone/zones cannot be specified as node's zone may conflict
+			if volumeOptions.ZonePresent || volumeOptions.ZonesPresent {
+				return nil, fmt.Errorf("Zone/Zones cannot be specified in StorageClass if VolumeBindingMode set to WaitForFirstConsumer.")
+			}
 
-	name, err := cloud.CreateDisk(volumeOptions)
-	if err != nil {
-		glog.V(2).Infof("Error creating EBS Disk volume: %v", err)
-		return "", 0, nil, "", err
-	}
-	glog.V(2).Infof("Successfully created EBS Disk volume %s", name)
+			// if node is set, we pick node's zone and ignore any allowedTopologies (since scheduler has accounted for it already)
+			if node.ObjectMeta.Labels == nil {
+				return nil, fmt.Errorf("labels for node missing")
+			}
 
-	labels, err := cloud.GetVolumeLabels(name)
-	if err != nil {
-		// We don't really want to leak the volume here...
-		glog.Errorf("error building labels for new EBS volume %q: %v", name, err)
+			z, ok := node.ObjectMeta.Labels[kubeletapis.LabelZoneFailureDomain]
+			if !ok {
+				return nil, fmt.Errorf("%s Label for node missing", kubeletapis.LabelZoneFailureDomain)
+			}
+			volumeOptions.ZonePresent = true
+			volumeOptions.AvailabilityZone = z
+		} else {
+			// node not specified so collect all specified zones from allowedTopologies
+			allowedZones := make([]string, 0)
+			for _, term := range allowedTopologies {
+				zoneRequirementFound := false
+				for _, exp := range term.MatchLabelExpressions {
+					if exp.Key == kubeletapis.LabelZoneFailureDomain {
```

Comment:
I wonder if this should error anything other than zone is specified.  Because that is information the user expects us to use, but we didn't.  cc @verult

Reply:
I guess they could be arbitrary labels. The example in https://github.com/kubernetes/community/blob/master/contributors/design-proposals/storage/volume-topology-scheduling.md#multi-label-volume refers to `key: foo.io/rack` which should be okay?

Reply:
Topology keys are a little different than label keys in Kubernetes, in that the topology keys are defined by the storage plugin and not the user, and the topology labels define the inherent restrictions of the storage provider.  So I think we shouldn't mix plugin and user-defined keys in allowedTopologies.  I think the safest course of action now is to error if there's a key we don't recognize.